Unable to update device (adafruit feather esp32-s2) on ESPhome 2025.10

### The problem

Unable to perform update/install, when using ESPHome 2025.10.3 or 2025.10.4.

Stuck using Arduino framework, as I'm using a custom component for a Adafruit NeoKey 1x4 QT I2C Breakout.


### Which version of ESPHome has the issue?

2025.10.4

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP32

### Component causing the issue

_No response_

### YAML Config

```yaml
substitutions:
  device_name: adaesp32
  friendly_name: AdaESP32

esphome:
  name: ${device_name}
  friendly_name: ${friendly_name}
  build_path: ./${device_name}_build/


esp32:
  board: featheresp32-s2
  #variant: esp32s2
  framework:
    type: arduino
#    type: esp-idf

logger:
  level: DEBUG

api:
  encryption:
    key: !secret api_key

ota:
  - platform: esphome
    password: !secret ota_password
    
wifi:
  power_save_mode: LIGHT
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  fast_connect: true
  manual_ip:
    static_ip: 192.168.10.11
    gateway: 192.168.10.1
    subnet: 255.255.255.0
```

### Anything in the logs that might be useful for us?

```txt
HARDWARE: ESP32S2 240MHz, 320KB RAM, 4MB Flash
 - framework-arduinoespressif32 @ 3.2.1 
 - framework-arduinoespressif32-libs @ 5.4.0+sha.858a988d6e 
 - framework-espidf @ 3.50402.0 (5.4.2) 
 - tool-cmake @ 3.30.2 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.0.2 
 - tool-mklittlefs @ 3.2.0 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20241119
Reading CMake configuration...
Dependency Graph
|-- Networking @ 3.2.1
|-- ESPmDNS @ 3.2.1
|-- Update @ 3.2.1
|-- Wire @ 3.2.1
|-- noise-c @ 0.1.10
Compiling .pioenvs/adaesp32/src/main.cpp.o
Compiling .pioenvs/adaesp32/chip-debug-report.cpp.o
Compiling .pioenvs/adaesp32/esp32-hal-misc.c.o
Compiling .pioenvs/adaesp32/esp32-hal-periman.c.o
Compiling .pioenvs/adaesp32/esp32-hal-psram.c.o
Compiling .pioenvs/adaesp32/esp32-hal-rgb-led.c.o
Compiling .pioenvs/adaesp32/esp32-hal-sigmadelta.c.o
Compiling .pioenvs/adaesp32/esp32-hal-spi.c.o
Compiling .pioenvs/adaesp32/esp32-hal-time.c.o
Compiling .pioenvs/adaesp32/esp32-hal-timer.c.o
In file included from /data/cache/platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:202,
                 from /data/cache/platformio/packages/framework-arduinoespressif32/cores/esp32/chip-debug-report.cpp:17:
/data/cache/platformio/packages/framework-arduinoespressif32/cores/esp32/chip-debug-report.cpp: In function 'void printBeforeSetupInfo()':
/data/cache/platformio/packages/framework-arduinoespressif32/cores/esp32/HardwareSerial.h:416:16: error: 'USBSerial' was not declared in this scope; did you mean 'Serial'?
  416 | #define Serial USBSerial
      |                ^~~~~~~~~
/data/cache/platformio/packages/framework-arduinoespressif32/cores/esp32/chip-debug-report.cpp:294:3: note: in expansion of macro 'Serial'
  294 |   Serial.begin(0);
      |   ^~~~~~
*** [.pioenvs/adaesp32/chip-debug-report.cpp.o] Error 1
========================== [FAILED] Took 5.16 seconds ==========================
```

### Additional information

I've tried to swap from 

```
esp32:  
  board: featheresp32-s2
```

To 

```
esp32:  
  variant: esp32s2
```

But get an error:

```  
  This pin cannot be used on ESP32-S2s and is already used by the SPI/PSRAM interface (function: SPID).
  pin: 
    number: A4
    allow_other_uses: True
```

If i try to add "ignore_pin_validation_error: true", to this, I get a different error:

```
 Only pins 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21 support wakeup.
  wakeup_pin: 
    number: A4
    ignore_pin_validation_error: True
    allow_other_uses: True
    inverted: False
    mode: 
      input: True
      pullup: True
```

--- a/esphome/components/esp32/__init__.py
+++ b/esphome/components/esp32/__init__.py
@@ -931,6 +931,12 @@ async def to_code(config):
         add_idf_sdkconfig_option("CONFIG_MBEDTLS_CERTIFICATE_BUNDLE", True)
         add_idf_sdkconfig_option("CONFIG_ESP_PHY_REDUCE_TX_POWER", True)
 
+        # ESP32-S2 Arduino: Disable USB Serial on boot to avoid TinyUSB dependency
+        if get_esp32_variant() == VARIANT_ESP32S2:
+            cg.add_build_unflag("-DARDUINO_USB_CDC_ON_BOOT=1")
+            cg.add_build_unflag("-DARDUINO_USB_CDC_ON_BOOT=0")
+            cg.add_build_flag("-DARDUINO_USB_CDC_ON_BOOT=0")
+
     cg.add_build_flag("-Wno-nonnull-compare")
 
     add_idf_sdkconfig_option(f"CONFIG_IDF_TARGET_{variant}", True)

